tests/mnpe_test.py

```diff
@@ -0,0 +1,282 @@
+        embedding_net=x_embedding if use_embed_net else torch.nn.Identity(),
+        log_transform_x=False,
+    )
+    trainer = MNPE(density_estimator=density_estimator)
+    trainer.append_simulations(theta, x).train(max_num_epochs=1)
+
+    # Test different samplers
+    posterior = trainer.build_posterior(prior=prior)
+    posterior.set_default_x(x_o)
+    if isinstance(posterior, VIPosterior):
+        posterior.train().sample((1,))
+    elif isinstance(posterior, RejectionPosterior):
+        posterior.sample((1,))
+    else:
+        posterior.sample((1,))
+
+
+class MixedSimulator(torch.nn.Module):
+    def __init__(self, sigma2):
+        super().__init__()
+        self.sigma2 = sigma2
+
+    def forward(self, theta):
+        # return x~N(mu_z, tau^2)
+        mu = theta[:, :2]
+        z = theta[:, -1]
+        # Select mu based on z
+        mu_z = torch.where(z == 0, mu[:, 0], mu[:, 1])
+        # Sample from N(mu_z, sigma2)
+        x = Normal(loc=mu_z, scale=torch.sqrt(self.sigma2)).rsample()
+        return x.reshape(-1, 1)
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize("flow_model", ("nsf", "zuko_nsf"))
+def test_mnpe_accuracy(
+    flow_model: str,
+):
+    """Test MNPE accuracy."""
+    num_simulations = 1000
+    num_posterior_samples = 1000
+
+    z_score_theta = "none"
+    use_embed_net = False
+
+    # Prio
+    mu_0 = torch.tensor([-1.0])
+    mu_1 = torch.tensor([1.0])
+    tau2 = torch.tensor([0.1])
+    prior_mu0 = Normal(loc=mu_0, scale=torch.sqrt(tau2))
+    prior_mu1 = Normal(loc=mu_1, scale=torch.sqrt(tau2))
+
+    q = torch.tensor(0.8)
+    prior_z = Bernoulli(probs=q * torch.ones(1))
+
+    prior = MultipleIndependent(
+        [prior_mu0, prior_mu1, prior_z],
+    )
+
+    # Simulator
+    sigma2 = torch.tensor(0.1)
+    simulator = MixedSimulator(sigma2=sigma2)
+
+    # observation
+    theta_o = prior.sample((1,))
+    x_o = simulator(theta_o)
+
+    ############################################################
+    # MNPE
+    num_simulations = 4000
+    num_posterior_samples = 1_000
+
+    theta = prior.sample((num_simulations,))
+    x = simulator(theta)
+
+    # Build estimator manually
+    x_embedding = FCEmbedding(1, 1)  # simple embedding net, 1 continuous parameter
+    density_estimator = posterior_nn(
+        model="mnpe",
+        flow_model=flow_model,
+        z_score_theta=z_score_theta,
+        embedding_net=x_embedding if use_embed_net else torch.nn.Identity(),
+        log_transform_x=False,
+    )
+    trainer = MNPE(density_estimator=density_estimator)
+    trainer.append_simulations(theta, x).train()
+
+    # Test different samplers
+    posterior_mnpe = trainer.build_posterior(prior=prior)
+    posterior_mnpe.set_default_x(x_o)
+
+    posterior_mnpe_samples = posterior_mnpe.sample((num_posterior_samples,))
+
+    ############################################################
+    # True posterior
+    ###### Posterior of z ######
+    s2 = tau2 + sigma2
```

Comment:
I think this should live in a separate function, we might want to re-use it somewhere else and the test would become more readable.